### DISCUSSION TITLE
When should a card be considered "known" for purposes of skipping?

### DISCUSSION THREAD

**Discussion by xofm31** [mortii/anki-morphs#178] [General]
When I finish my reviews on the computer for the day, I have most new cards in the "learning" state, which I will review again later on my phone. When I run Ankimorph recalc at that point, cards that are in the learning state stay in the new queue, and are still listed as "unknown" in the am-highlighted field.  I believe the relevant line in recalc.py is `if morph.highest_learning_interval == 0:`.

Personally, I feel like once I have seen the morph, even if it has  not gotten to a day or more learning interval, I would want it to be considered as no longer new, and to be moved to the default difficulty, and the am-highlighted field to mark it as "learning". 

What do you think @mortii ? I believe this is the way Morphman works, but in looking at the Morphman code for a couple of minutes, I didn't see where it calculates this.

**Comment by mortii** [comment]
I didn't anticipate anyone using recalc while cards are still in the learning state, so it wouldn't surprise me if it leads to weird/unexpected results. 

> Personally, I feel like once I have seen the morph, even if it has not gotten to a day or more learning interval, I would want it to be considered as no longer new, and to be moved to the default difficulty, [...]

Do you mean that other new cards that have the same 'learning' focus morph be moved to the default difficulty position, i.e. the back of the queue?

> [...] and the am-highlighted field to mark it as "learning".

That's completely reasonable. The caching process is very convoluted, so I'm not sure off the top of my head how much rewriting that would require, but hopefully we can just add the equivalent of  `if card.queue == 'learning'` somewhere.

**Comment by xofm31** [comment]
> Do you mean that other new cards that have the same 'learning' focus morph be moved to the default difficulty position, i.e. the back of the queue?

Yes, that's right, because once I've chosen which card with that morph is the one I'm going to learn, I want it to be treated as having been seen, regardless of whether I've learned it or not.

>  but hopefully we can just add the equivalent of if card.queue == 'learning' somewhere

Unfortunately I don't see a way to do this in recalc.py, because this is a property of the morph, and not of the card whose difficulty is being calculated. It looks like the only information that we have on the morph is the `highest_learning_interval`, which is an integer in the database.

`        SELECT Card_Morph_Map.card_id, Morphs.lemma, Morphs.inflection, Morphs.highest_learning_interval`

I don't see where `Morphs.highest_learning_interval` gets calculated and stored in the database initially. Perhaps if unseen morphs had a value of -1, then the ones in the learning queue with an interval of less than a day could easily be distinguished, and it would be a matter of changing to `if morph.highest_learning_interval == -1:`?

**Reply by xofm31** [reply]
Or perhaps when rounding the `highest_learning_interval`, you could round up, so that even if the interval is 15 min, it would show up as 1 day. This might require changing the logic for when it becomes known, though...

**Comment by mortii** [comment]
The intervals (`card.ivl`) get extracted from here:

https://github.com/mortii/anki-morphs/blob/6e3d3be6eec295e457d91cc4a05326880a31dabe/ankimorphs/recalc.py#L345-L353

If I remember correctly the intervals for learning cards are 0, so no rounding is taking place. 

The `highest_interval` is set here:

https://github.com/mortii/anki-morphs/blob/6e3d3be6eec295e457d91cc4a05326880a31dabe/ankimorphs/recalc.py#L196-L199

That is where I imagined we could add the additional if statement. That could probably work with your idea of the default being -1 for new cards. I'm pretty sure there are many hardcoded statements for like this:`highest_interval > 0` throughout the code, so some legwork would be required to change all of those, but it might be the best approach.

**Comment by xofm31** [comment]
It looks like it's just a matter of checking for the card type, and then setting the highest_interval to 1 if it's a card in the learning state.

https://github.com/xofm31/anki-morphs/commit/5930f988ac26744ee2097cf3fffb6840a0bc262a

I couldn't run precommit; venv and conda seem to be fighting with each other.

**Reply by mortii** [reply]
Beautiful! Create a pull request and I'll merge it after v2 is complete. 

The test suite in pre-commit will fail  (most changes to recalc does), but I can easily fix that later.

**Comment by mortii** [comment]
Released in [v1.3.0](https://github.com/mortii/anki-morphs/releases/tag/v1.3.0)!
